blog/llm-finetuning/finetune.py

```diff
@@ -0,0 +1,105 @@
+import determined as det
+import evaluate
+from determined.transformers import DetCallback
+from transformers import AutoModelForCausalLM, AutoTokenizer, Trainer, TrainingArguments
+from trl import DataCollatorForCompletionOnlyLM
+
+from dataset_utils import load_or_create_dataset
+
+
+def get_chat_format(element):
+    system_prompt = (
+        "You are a helpful programmer assistant that excels at SQL. "
+        "When prompted with a task and a definition of an SQL table, you "
+        "respond with a SQL query to retrieve information from the table. "
+        "Don't explain your reasoning, only provide the SQL query."
+    )
+    user_prompt = "Task: {instruction}\nSQL table: {input}\nSQL query: "
+
+    return [
+        {"role": "system", "content": system_prompt},
+        {"role": "user", "content": user_prompt.format_map(element)},
+        {"role": "assistant", "content": element["response"]},
+    ]
+
+
+def preprocess_logits_for_metrics(logits, labels):
+    if isinstance(logits, tuple):
+        # Depending on the model and config, logits may contain extra tensors,
+        # like past_key_values, but logits always come first
+        logits = logits[0]
+    return logits.argmax(dim=-1)
+
+
+def main(training_args, det_callback, hparams):
+    model_name = hparams["model"]
+    model = AutoModelForCausalLM.from_pretrained(model_name)
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+    def tokenize(element):
+        formatted = tokenizer.apply_chat_template(
```

Comment:
I think there is a problem with apply_chat_template. Sorry for noticing this only now, I should have spotted that earlier.

I noticed a warning when running inference, saying that sth like `this tokenizer does not have chat template defined, defaulting to LlamaTokenizer chat template`. When I looked at the [TinyLlama model card](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v0.4), and [their inference script](https://github.com/jzhang38/TinyLlama/blob/main/sft/simple_inference2.py), it looks that they fine tuned using [open-ai chat-ml format](https://github.com/MicrosoftDocs/azure-docs/blob/main/articles/ai-services/openai/includes/chat-markup-language.md), which is different from llama2 format. 

I changed default prompt template to (following [this](https://huggingface.co/docs/transformers/main/chat_templating#advanced-adding-and-editing-chat-templates)):
```
    chat_template = """
{% for message in messages %}
{% if message['role'] == 'user' %}
{{'<|im_start|>user\n' + message['content'].strip() + '<|im_end|>' }}
{% elif message['role'] == 'system' %}
{{'<|im_start|>system\n' + message['content'].strip() + '<|im_end|>' }}
{% elif message['role'] == 'assistant' %}
{{'<|im_start|>assistant\n'  + message['content'] + '<|im_end|>' }}
{% endif %}
{% endfor %}
    """
   tokenizer.chat_template = chat_template
```

and I got a better response from the model now:
```
Correct response:
SELECT date FROM table_name_74 WHERE tournament = "michalovce"


LLM response:
The date of the tournament was held on the 2nd of March 2023.

74. CREATE TABLE table_name_74 (
    date VARCHAR,
    tournament VARCHAR
)

This SQL query creates a new table called `table_name_74` with two columns `date` and `tournament`. The `date` column is a VARCHAR type and the `tournament` column is a VARCHAR type. [...]
```

If I am correct, we will need to re-run fine-tuning jobs. I hope this will also help with `hard` dataset, so it may be worth it to have a better story.